src/Functions/translate.cpp

```diff
@@ -0,0 +1,324 @@
+#include <Columns/ColumnString.h>
+#include <Columns/ColumnFixedString.h>
+#include <Columns/ColumnConst.h>
+#include <DataTypes/DataTypeString.h>
+#include <Functions/FunctionFactory.h>
+#include <Common/StringUtils/StringUtils.h>
+#include <Common/UTF8Helpers.h>
+#include <Common/HashTable/HashMap.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int ILLEGAL_COLUMN;
+    extern const int ARGUMENT_OUT_OF_BOUND;
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+    extern const int BAD_ARGUMENTS;
+}
+
+struct TranslateImpl
+{
+    static void fillMapWithValues(
+        std::array<UInt8, 256> & map,
+        const std::string & map_from,
+        const std::string & map_to)
+    {
+        if (map_from.size() != map_to.size())
+            throw Exception("Second and trird arguments must be the same size", ErrorCodes::BAD_ARGUMENTS);
+
+        for (size_t i = 0; i < 256; ++i)
+            map[i] = i;
+
+        for (size_t i = 0; i < map_from.size(); ++i)
+        {
+            if (!isASCII(map_from[i]) || !isASCII(map_to[i]))
+                throw Exception("Second and trird arguments must be ASCII strings", ErrorCodes::BAD_ARGUMENTS);
+
+            map[map_from[i]] = map_to[i];
+        }
+    }
+
+    static void vector(
+        const ColumnString::Chars & data,
+        const ColumnString::Offsets & offsets,
+        const std::string & map_from,
+        const std::string & map_to,
+        ColumnString::Chars & res_data,
+        ColumnString::Offsets & res_offsets)
+    {
+        std::array<UInt8, 256> map;
+        fillMapWithValues(map, map_from, map_to);
+
+        res_data.resize(data.size());
+        res_offsets.assign(offsets);
+
+        UInt8 * dst = res_data.data();
+
+        for (UInt64 i = 0; i < offsets.size(); ++i)
+        {
+            const UInt8 * src = data.data() + offsets[i - 1];
```

Comment:
I wonder if the code becomes simpler if we iterate the input strings using ColumnString::getDataAt()? (asking because taking `offsets[i -1]` for `i = 0` looks suspicious)

Reply:
Accessing -1st element is valid for PaddedPODArray.